amm/contracts/router/lib.rs

```diff
@@ -653,6 +662,59 @@ pub mod router {
         ) -> Result<Vec<u128>, RouterError> {
             self.calculate_amounts_in(amount_out, &path)
         }
+
+        /// Caches the pair address for the given pair.
+        #[ink(message)]
+        fn cache_pair(
```

Comment:
I think this method should not allow `pair_address` parameter, but rather call the factory to get it. That strategy keeps the previous invariant intact (that both router and factory see the same "pair universe")

Reply:
My idea was that there could possibly be multiple pairs for the same `(token0, token1)` pair but that's not true - `Factory` checks whether it already has the pair created. 

It's actually worse, I need to rethink my approach (actually I already had started and abandoned it).